# Import Libraries

In [2]:
# Importing libraries
import pandas as pd
import numpy as np
import os

## Create a path for folder

In [5]:
# Turn the data folder path into string
path = r'/Users/amritkaur/Desktop/CF Data Immersion/Chapter 4/Achievement 4'

## Import Data from Prepared data folder

In [10]:
# Import data from Prepared Data folder
df_ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_derived.pkl'))

In [12]:
df_ords_prods_merge.shape

(32404859, 18)

# Grouping and Aggregating Data on entire dataframe

In [17]:
# grouping 'product_name' column using groupby() function
df_ords_prods_merge.groupby('product_name')

In [21]:
# To calculate the mean of the “order_number” column grouped by the “department_id” column.
df_ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

### How do the results for the entire dataframe differ from those of the subset? 
The subset of dataframe contains only certain rows or department IDs, and it represents a specific portion of the data. In this case, for department ID 4, the mean value is higher in the subset (18.825780). In entire dataframe, when we expand the analysis to include the full dataset, the mean value for department ID 4 decreases (17.811403). This suggests that when we incorporate the full data, the values from other department IDs or rows not included in the subset influence the overall mean.
If the full dataset includes some extreme values or outliers that were absent from the subset, these outliers could be distorting the mean in the entire dataframe. For example, if there are some very low values in department 4 that were not in the subset, the overall mean will decrease. If the subset you selected happened to be more representative of higher values for department 4, the mean in the subset could have been biased upward. 

# Create a Loyalty flag

In [28]:
# Aggregating Data with transform() Function
df_ords_prods_merge['max_order'] = df_ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

/var/folders/b2/nhtg2py55bd6x9ky11p928r00000gn/T/ipykernel_2650/3473307984.py:2: FutureWarning: The provided callable <function max at 0x103bf7ce0> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  df_ords_prods_merge['max_order'] = df_ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)


In [30]:
# Checking the Output
df_ords_prods_merge.head(15)

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0    2539329        1             1                  2                  8   
1    2539329        1             1                  2                  8   
2    2539329        1             1                  2                  8   
3    2539329        1             1                  2                  8   
4    2539329        1             1                  2                  8   
5    2398795        1             2                  3                  7   
6    2398795        1             2                  3                  7   
7    2398795        1             2                  3                  7   
8    2398795        1             2                  3                  7   
9    2398795        1             2                  3                  7   
10   2398795        1             2                  3                  7   
11    473747        1             3                  3                 12   
12    473747        1             3                  3                 12   
13    473747        1             3                  3                 12   
14    473747        1             3                  3                 12   

    days_since_last_order  product_id  add_to_cart_order  reordered _merge  \
0                     NaN         196                  1          0   both   
1                     NaN       14084                  2          0   both   
2                     NaN       12427                  3          0   both   
3                     NaN       26088                  4          0   both   
4                     NaN       26405                  5          0   both   
5                    15.0         196                  1          1   both   
6                    15.0       10258                  2          0   both   
7                    15.0       12427                  3          1   both   
8                    15.0       13176                  4          0   both   
9                    15.0       26088                  5          1   both   
10                   15.0       13032                  6          0   both   
11                   21.0         196                  1          1   both   
12                   21.0       12427                  2          1   both   
13                   21.0       10258                  3          1   both   
14                   21.0       25133                  4          0   both   

                               product_name  aisle_id  department_id  prices  \
0                                      Soda        77              7     9.0   
1   Organic Unsweetened Vanilla Almond Milk        91             16    12.5   
2                       Original Beef Jerky        23             19     4.4   
3                Aged White Cheddar Popcorn        23             19     4.7   
4          XL Pick-A-Size Paper Towel Rolls        54             17     1.0   
5                                      Soda        77              7     9.0   
6                                Pistachios       117             19     3.0   
7                       Original Beef Jerky        23             19     4.4   
8                    Bag of Organic Bananas        24              4    10.3   
9                Aged White Cheddar Popcorn        23             19     4.7   
10                    Cinnamon Toast Crunch       121             14     4.0   
11                                     Soda        77              7     9.0   
12                      Original Beef Jerky        23             19     4.4   
13                               Pistachios       117             19     3.0   
14                    Organic String Cheese        21             16     8.6   

   match     Busiest Day    Busiest_days busiest_period_of_day  max_order  
0   both  Regularly busy  Regularly busy        Average orders         10  
1   both  Regularly busy  Regularly busy        Average orders         10  
2   both  Reg

In [34]:
# Creating a 'Loyalty' Label Flag using loc Function
df_ords_prods_merge.loc[df_ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [36]:
df_ords_prods_merge.loc[(df_ords_prods_merge['max_order'] <= 40) & (df_ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [38]:
df_ords_prods_merge.loc[df_ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [40]:
# Examining Value Counts on 'loyalty_flag'
df_ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

In [42]:
# Checking the Updated Output
df_ords_prods_merge.head(15)

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0    2539329        1             1                  2                  8   
1    2539329        1             1                  2                  8   
2    2539329        1             1                  2                  8   
3    2539329        1             1                  2                  8   
4    2539329        1             1                  2                  8   
5    2398795        1             2                  3                  7   
6    2398795        1             2                  3                  7   
7    2398795        1             2                  3                  7   
8    2398795        1             2                  3                  7   
9    2398795        1             2                  3                  7   
10   2398795        1             2                  3                  7   
11    473747        1             3                  3                 12   
12    473747        1             3                  3                 12   
13    473747        1             3                  3                 12   
14    473747        1             3                  3                 12   

    days_since_last_order  product_id  add_to_cart_order  reordered _merge  \
0                     NaN         196                  1          0   both   
1                     NaN       14084                  2          0   both   
2                     NaN       12427                  3          0   both   
3                     NaN       26088                  4          0   both   
4                     NaN       26405                  5          0   both   
5                    15.0         196                  1          1   both   
6                    15.0       10258                  2          0   both   
7                    15.0       12427                  3          1   both   
8                    15.0       13176                  4          0   both   
9                    15.0       26088                  5          1   both   
10                   15.0       13032                  6          0   both   
11                   21.0         196                  1          1   both   
12                   21.0       12427                  2          1   both   
13                   21.0       10258                  3          1   both   
14                   21.0       25133                  4          0   both   

                               product_name  aisle_id  department_id  prices  \
0                                      Soda        77              7     9.0   
1   Organic Unsweetened Vanilla Almond Milk        91             16    12.5   
2                       Original Beef Jerky        23             19     4.4   
3                Aged White Cheddar Popcorn        23             19     4.7   
4          XL Pick-A-Size Paper Towel Rolls        54             17     1.0   
5                                      Soda        77              7     9.0   
6                                Pistachios       117             19     3.0   
7                       Original Beef Jerky        23             19     4.4   
8                    Bag of Organic Bananas        24              4    10.3   
9                Aged White Cheddar Popcorn        23             19     4.7   
10                    Cinnamon Toast Crunch       121             14     4.0   
11                                     Soda        77              7     9.0   
12                      Original Beef Jerky        23             19     4.4   
13                               Pistachios       117             19     3.0   
14                    Organic String Cheese        21             16     8.6   

   match     Busiest Day    Busiest_days busiest_period_of_day  max_order  \
0   both  Regularly busy  Regularly busy        Average orders         10   
1   both  Regularly busy  Regularly busy        Average orders         10   
2   both  

# Analyze the spending habits of Loyalty customers

In [45]:
# Determining Basic Statistics of Spending Habits by Loyalty Category
df_ords_prods_merge.groupby('loyalty_flag').agg({'prices' : ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

As we can see that Loyal customers generally purchases low-cost items with the average of $10.39 and while new customers tend to spend more, averaging $13.29 per item. Regular customers fall between the two, with an average cost of $12.50 per item. 
While all customer groups have a minimum price of $1.00, the presence of an unusually high maximum price of $99,999 indicates that there may be outliers or data errors skewing the results.

# Create a Spending flag

In [54]:
# Aggregating Data with transform() Function
df_ords_prods_merge['avg_product_price'] = df_ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

/var/folders/b2/nhtg2py55bd6x9ky11p928r00000gn/T/ipykernel_2650/1144848887.py:2: FutureWarning: The provided callable <function mean at 0x103c10720> is currently using SeriesGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  df_ords_prods_merge['avg_product_price'] = df_ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)


In [64]:
# Checking Relevant Columns 
df_ords_prods_merge[['user_id', 'avg_product_price']].head(100)

user_id  avg_product_price
0         1           6.367797
1         1           6.367797
2         1           6.367797
3         1           6.367797
4         1           6.367797
..      ...                ...
95        2           7.515897
96        2           7.515897
97        2           7.515897
98        2           7.515897
99        2           7.515897

[100 rows x 2 columns]

In [66]:
# Creating a 'Spending' Label Flag using loc Function
df_ords_prods_merge.loc[df_ords_prods_merge['avg_product_price'] >= 10, 'spending_flag'] = 'High spender'
df_ords_prods_merge.loc[df_ords_prods_merge['avg_product_price'] < 10, 'spending_flag'] = 'Low spender'

In [68]:
# Examining Value Count on 'spending_flag' Column
df_ords_prods_merge['spending_flag'].value_counts(dropna = False)

spending_flag
Low spender     31770614
High spender      634245
Name: count, dtype: int64

In [99]:
# Checking Releavent Columns
df_ords_prods_merge[['user_id', 'avg_product_price', 'spending_flag']].head(100)

user_id  avg_product_price spending_flag
0         1           6.367797   Low spender
1         1           6.367797   Low spender
2         1           6.367797   Low spender
3         1           6.367797   Low spender
4         1           6.367797   Low spender
..      ...                ...           ...
95        2           7.515897   Low spender
96        2           7.515897   Low spender
97        2           7.515897   Low spender
98        2           7.515897   Low spender
99        2           7.515897   Low spender

[100 rows x 3 columns]

# Create an order frequency flag

In [75]:
# Aggregating Data with transform() Function
df_ords_prods_merge['median_days_between_orders'] = df_ords_prods_merge.groupby(['user_id'])['days_since_last_order'].transform(np.median)

/var/folders/b2/nhtg2py55bd6x9ky11p928r00000gn/T/ipykernel_2650/3603756689.py:2: FutureWarning: The provided callable <function median at 0x103d2b420> is currently using SeriesGroupBy.median. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "median" instead.
  df_ords_prods_merge['median_days_between_orders'] = df_ords_prods_merge.groupby(['user_id'])['days_since_last_order'].transform(np.median)


In [101]:
# Checking Relevant Columns 
df_ords_prods_merge[['user_id', 'days_since_last_order', 'median_days_between_orders']].head(100)

user_id  days_since_last_order  median_days_between_orders
0         1                    NaN                        20.5
1         1                    NaN                        20.5
2         1                    NaN                        20.5
3         1                    NaN                        20.5
4         1                    NaN                        20.5
..      ...                    ...                         ...
95        2                    8.0                        13.0
96        2                    8.0                        13.0
97        2                    8.0                        13.0
98        2                    8.0                        13.0
99        2                    8.0                        13.0

[100 rows x 3 columns]

In [79]:
# Creating a 'Order Frequency' Label Flag using loc Function
df_ords_prods_merge.loc[df_ords_prods_merge['median_days_between_orders'] > 20, 'order_frequency_flag'] = 'Non-frequent customer'

In [81]:
df_ords_prods_merge.loc[(df_ords_prods_merge['median_days_between_orders'] <= 20) & (df_ords_prods_merge['median_days_between_orders'] > 10), 'order_frequency_flag'] = 'Regular customer'

In [83]:
df_ords_prods_merge.loc[df_ords_prods_merge['median_days_between_orders'] <= 10, 'order_frequency_flag'] = 'Frequent customer'

In [85]:
# Examining Value Counts on 'order_frequency_flag'
df_ords_prods_merge['order_frequency_flag'].value_counts(dropna = False)

order_frequency_flag
Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: count, dtype: int64

In [93]:
# Checking Relevant Columns of Updated Dataframe
df_ords_prods_merge[['user_id', 'days_since_last_order', 'median_days_between_orders', 'order_frequency_flag']].head(100)

user_id  days_since_last_order  median_days_between_orders  \
0         1                    NaN                        20.5   
1         1                    NaN                        20.5   
2         1                    NaN                        20.5   
3         1                    NaN                        20.5   
4         1                    NaN                        20.5   
..      ...                    ...                         ...   
95        2                    8.0                        13.0   
96        2                    8.0                        13.0   
97        2                    8.0                        13.0   
98        2                    8.0                        13.0   
99        2                    8.0                        13.0   

     order_frequency_flag  
0   Non-frequent customer  
1   Non-frequent customer  
2   Non-frequent customer  
3   Non-frequent customer  
4   Non-frequent customer  
..                    ...  
95       Regular customer  
96       Regular customer  
97       Regular customer  
98       Regular customer  
99       Regular customer  

[100 rows x 4 columns]

# Export the Dataframe in pickle Format to the "Prepared Data" Folder

In [96]:
# Exporting the  updated dataframe
df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_grouped.pkl'))